### PR TITLE
Add CloudService UI

### LINKED ISSUE #5872: Cloud Service Model
As a network engineer, I want to be able to represent generic cloud services for components that do not have a concrete data model available, to be able to properly represent cloud deployments.

Fields:
- name (unique)
- type: FK to cloud type
- cloud network: FK to cloud network
- account: FK to cloud account (optional)
- extra config: JSON field

### PR DESCRIPTION
**Pull request by timizuoebideri1**
# Closes #5872 
# What's Changed

# Screenshots

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example App Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/cloud/forms.py [hunk 1]
```diff
@@ -136,3 +136,66 @@ class CloudTypeFilterForm(NautobotFilterForm):
...
+    )
+
+    class Meta:
+        model = CloudService
+        fields = [
+            "name",
+            "cloud_account",
+            "cloud_network",
+            "cloud_type",
+            "extra_config",
```
[the remark this answers is not in the document]
Comment (timizuoebideri1) [reply nautobot/cloud/forms.py:164]: Screenshot 2024-07-09 at 5 09 22 PM

File: nautobot/cloud/forms.py [hunk 2]
```diff
@@ -136,3 +136,63 @@ class CloudTypeFilterForm(NautobotFilterForm):
...
+
+
+#
+# Cloud Service
+#
+
+
+class CloudServiceForm(NautobotModelForm):
+    cloud_account = DynamicModelChoiceField(
+        queryset=CloudAccount.objects.all(),
```
Comment (glennmatthews) [inline nautobot/cloud/forms.py:148]: Does this need to be `required=False`?

File: nautobot/cloud/navigation.py [hunk 3]
```diff
@@ -46,6 +46,22 @@
                             ),
                         ),
                     ),
+                    NavMenuItem(
```
Comment (glennmatthews) [inline nautobot/cloud/navigation.py:49]: Let's think carefully about ordering of the Cloud menu items. To match the other menus, should `Cloud Services` be located above `Cloud Types` at least?

File: nautobot/cloud/templates/cloud/cloudservice_retrieve.html [hunk 4]
```diff
@@ -0,0 +1,45 @@
+{% extends 'generic/object_retrieve.html' %}
+{% load helpers %}
+
+{% block extra_breadcrumbs %}
+    <li><a href="{% url 'cloud:cloudservice_list' %}?cloud_network={{ object.cloud_network.pk }}">{{ object.cloud_network }}</a></li>
```
Comment (glennmatthews) [inline nautobot/cloud/templates/cloud/cloudservice_retrieve.html:5]: This seems strange to me? Is it correct per our data model?

File: nautobot/cloud/templates/cloud/cloudservice_retrieve.html [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    </div>
+{% endblock content_left_page %}
+
+{% block content_full_width_page %}
+    <div class="panel panel-default">
+        <div class="panel-heading">
+            <strong>Extra Config</strong>
+        </div>
+        <div class="panel-body">
+            <pre>{{ object.extra_config }}</pre>
```
Comment (glennmatthews) [inline nautobot/cloud/templates/cloud/cloudservice_retrieve.html:42]: We should use `render_json` here for syntax highlighting, shouldn't we?

File: nautobot/cloud/templates/cloud/cloudservice_retrieve.html [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+            </tr>
+            <tr>
+                <td>Cloud Type</td>
+                <td>{{ object.cloud_type|hyperlinked_object }}</td>
+            </tr>
+        </table>
+    </div>
+{% endblock content_left_page %}
+
+{% block content_full_width_page %}
```
Comment (glennmatthews) [inline nautobot/cloud/templates/cloud/cloudservice_retrieve.html:36]: *maybe* content_right_page instead?

File: nautobot/cloud/tests/test_views.py [hunk 7]
```diff
@@ -64,3 +64,36 @@ def setUpTestData(cls):
...
+        CloudService.objects.create(
+            name="Deletable Cloud Service 3", cloud_type=cloud_types[2], cloud_network=cloud_networks[2]
+        )
+
+        cls.form_data = {
+            "name": "New Cloud Service",
+            "cloud_type": cloud_types[1].pk,
+            "cloud_network": cloud_networks[1].pk,
+            "cloud_account": cloud_accounts[1].pk,
+            # "extra_config": {"status": "greetings", "role": 1},
```
Comment (glennmatthews) [inline nautobot/cloud/tests/test_views.py:92]: to be enabled?

**Comment by HanlinMiao** [comment]
You can fix the test by adding `is_metadata_associable = False` to the User class definiton.

File: nautobot/cloud/forms.py [hunk 8]
```diff
@@ -136,3 +136,66 @@ class CloudTypeFilterForm(NautobotFilterForm):
...
+            "tags",
+        ]
+
+
+class CloudServiceBulkEditForm(TagsBulkEditFormMixin, NautobotBulkEditForm):
+    pk = forms.ModelMultipleChoiceField(queryset=CloudType.objects.all(), widget=forms.MultipleHiddenInput)
+    cloud_account = DynamicModelChoiceField(queryset=CloudAccount.objects.all(), required=False)
+    cloud_network = DynamicModelChoiceField(queryset=CloudNetwork.objects.all(), required=False)
+    cloud_type = DynamicModelChoiceField(queryset=CloudType.objects.all(), required=False)
+
```
[the remark this answers is not in the document]
Comment (timizuoebideri1) [reply nautobot/cloud/forms.py:175]: I may be mistaken, but I haven't actually seen a JSON config field in any Bulk operation forms.

File: nautobot/cloud/tests/test_views.py [hunk 9]
```diff
@@ -64,3 +64,36 @@ def setUpTestData(cls):
...
+        cloud_accounts = CloudAccount.objects.all()
+        CloudService.objects.create(
+            name="Deletable Cloud Service 1", cloud_type=cloud_types[0], cloud_network=cloud_networks[0]
+        )
+        CloudService.objects.create(
+            name="Deletable Cloud Service 2", cloud_type=cloud_types[1], cloud_network=cloud_networks[1]
+        )
+        CloudService.objects.create(
+            name="Deletable Cloud Service 3", cloud_type=cloud_types[2], cloud_network=cloud_networks[2]
+        )
```
Comment (HanlinMiao) [inline nautobot/cloud/tests/test_views.py:85]:
```suggestion
        CloudService.objects.create(
            name="Deletable Cloud Service 3", cloud_type=cloud_types[2], cloud_network=cloud_networks[2], cloud_account=cloud_accounts[0],
        )
```

**Review by HanlinMiao — APPROVED** [review]

**Review by glennmatthews — COMMENTED** [review]
Do we have documentation for the CloudService model yet? If not, do we have an issue tracking that?

Also please be sure to address https://github.com/nautobot/nautobot/blob/next/nautobot/cloud/models.py#L189-L191

File: nautobot/cloud/forms.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (HanlinMiao) [reply nautobot/cloud/forms.py:175]: Take a look at `ExternalIntegrationBulkEditForm` and its `extra_config` field.

**Comment by HanlinMiao** [comment]
> Do we have documentation for the CloudService model yet? If not, do we have an issue tracking that?
> 
> Also please be sure to address https://github.com/nautobot/nautobot/blob/next/nautobot/cloud/models.py#L189-L191

Model documentation for CloudService: https://github.com/nautobot/nautobot/issues/5913

File: nautobot/cloud/navigation.py [hunk 10]
```diff
@@ -30,10 +30,26 @@
...
+                                    "cloud.add_cloudservice",
+                                ],
+                            ),
+                        ),
+                    ),
                     NavMenuItem(
                         link="cloud:cloudtype_list",
                         name="Cloud Types",
-                        weight=200,
+                        weight=250,
```
Comment (glennmatthews) [inline nautobot/cloud/navigation.py:52]: Let's stick to multiples of 100 for the cloud menu items unless there's a strong need not to. See also https://github.com/nautobot/nautobot/pull/5916/files#r1674067073

File: nautobot/cloud/tables.py [hunk 11]
```diff
@@ -128,3 +128,33 @@ class Meta(BaseTable.Meta):
...
             "actions",
         )
+
+
+class CloudServiceTable(BaseTable):
+    pk = ToggleColumn()
+    name = tables.Column(linkify=True)
+    cloud_account = tables.Column(linkify=True)
+    # TODO(timizuo): Uncomment after CloudNetwork UI is implemented
+    # cloud_network = tables.Column(linkify=True)
```
Comment (glennmatthews) [inline nautobot/cloud/tables.py:138]: Can this be enabled now?

File: nautobot/cloud/models.py [hunk 12]
```diff
@@ -164,7 +164,6 @@ def clean(self):
 class CloudNetworkPrefixAssignment(BaseModel):
     cloud_network = models.ForeignKey(CloudNetwork, on_delete=models.CASCADE, related_name="prefix_assignments")
     prefix = models.ForeignKey("ipam.Prefix", on_delete=models.CASCADE, related_name="cloud_network_assignments")
-    is_metadata_associable_model = False
```
Comment (glennmatthews) [inline nautobot/cloud/models.py:167]: Is this change correct/desirable?

**Review by glennmatthews — APPROVED** [review]

**Pull request merged**